src/components/notifications/NotificationList.jsx

```diff
@@ -62,6 +63,26 @@ export default function NotificationList() {
     }
   }, []);
 
+  const onViewableItemsChanged = useCallback(({viewableItems}) => {
+    const impressions = [];
+    const visibleNotifications = viewableItems.filter(it => it.isViewable).map(({item}) => item);
+
+    setSeenNotifications(prevState => {
+      visibleNotifications.forEach(visible => {
+        const exists = prevState.find(prev => visible.notificationId in prev);
+        if (!exists) impressions.push(visible);
+      });
+      return [
+        ...prevState,
+        ...visibleNotifications.map(visible => ({
+          [visible.notificationId]: visible.notificationId,
+        })),
+      ];
+    });
+    dispatch(logImpression(impressions));
```

Comment:
setState is executed asynchronously, so this might not behave as expected, most likely
It might be a good idea to manage 'seenNotifications' with useRef.

```js

const seenNotifications = useRef({}).current;

// ...

visible.forEach((msg) => {
  if (!seenNotifications[msg.messageId]) {
    seenNotifications[msg.messageId] = true;
    impressions.push(msg);
  }
});
```

Reply:
Good catch. I've updated it to useRef.